bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/SMSModemHandlerFactory.java

```diff
@@ -0,0 +1,105 @@
+ */
+@Component(configurationPid = "binding.smsmodem", service = ThingHandlerFactory.class)
+@NonNullByDefault
+public class SMSModemHandlerFactory extends BaseThingHandlerFactory {
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<>();
+    {
+        SUPPORTED_THING_TYPES_UIDS.add(SMSConversationHandler.SUPPORTED_THING_TYPES_UIDS);
+        SUPPORTED_THING_TYPES_UIDS.addAll(SMSModemBridgeHandler.SUPPORTED_THING_TYPES_UIDS);
+    }
+
+    private @NonNullByDefault({}) SerialPortManager serialPortManager;
+
+    @Reference
+    protected void setSerialPortManager(final SerialPortManager serialPortManager) {
+        this.serialPortManager = serialPortManager;
+    }
+
+    protected void unsetSerialPortManager(final SerialPortManager serialPortManager) {
+        this.serialPortManager = null;
+    }
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (SMSModemBridgeHandler.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
+            return new SMSModemBridgeHandler((Bridge) thing, serialPortManager);
+        } else if (SMSConversationHandler.SUPPORTED_THING_TYPES_UIDS.equals(thingTypeUID)) {
+            return new SMSConversationHandler(thing);
+        }
+
+        return null;
+    }
+
+    @Override
+    public @Nullable Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration,
+            @Nullable ThingUID thingUID, @Nullable ThingUID bridgeUID) {
+        if (SMSModemBridgeHandler.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
+            return super.createThing(thingTypeUID, configuration, thingUID, null);
+        }
+        if (SMSConversationHandler.SUPPORTED_THING_TYPES_UIDS.equals(thingTypeUID)) {
+            if (bridgeUID != null) {
+                ThingUID safethingUID = thingUID == null
+                        ? getSMSConversationUID(thingTypeUID,
+                                (String) configuration
+                                        .get(SMSModemBindingConstants.SMSCONVERSATION_PARAMETER_RECIPIENT),
+                                bridgeUID)
+                        : thingUID;
+                return super.createThing(thingTypeUID, configuration, safethingUID, bridgeUID);
+            } else {
+                throw new IllegalArgumentException("Cannot create a SMSConversation without a SMSModem bridge");
+            }
+
+        }
+        throw new IllegalArgumentException("The thing type " + thingTypeUID + " is not supported by the binding.");
+    }
+
+    public static ThingUID getSMSConversationUID(ThingTypeUID thingTypeUID, String recipient, ThingUID bridgeUID) {
+        return new ThingUID(thingTypeUID, recipient, bridgeUID.getId());
```

Comment:
The "getSMSConversationUID" method is called by the autodiscovery method, or from the createThing method if and only if the "thingUID" is null.
With this method, I want to give a default UID, IF NOT set by the user. If the user set the thing UID, it will use the user's one.
Your comment seems to mean that this approach is invalid, or do you mean it's badly implemented ?

Reply:
In `SMSConversationDiscoveryService` you are calling `getSMSConversationUID()` directly, when creating the discovered Thing (which is the right approach). So, the default-ThingUID-use-case is handled there.

If the user adds the Thing manually, the ThingUID can't be left empty/null (neither via UI nor text based configuration).

What use case do you intend to handle by this code?

Reply:
No particular use case here. I was just trying to implement the createThing method from the parent class and respecting the annotation : in this method, the parameter thingUID is nullable.
If you thing it is best, I can remove the call to the getSMSConversationUID in createThing, but the compiler will log a warning I guess ?
Or I could fail fast with an exception if thingUID is null.
What do you think ?

Reply:
Oh my ! I wouldn't have wasted your time if I knew that before... Sorry 😨
Deleted !